modules/basicfuncs/list-funcs.c

```diff
@@ -328,3 +329,305 @@ tf_implode(LogMessage *msg, gint argc, GString *argv[], GString *result)
+    default:
+      return TRUE;
+    }
+}
+
+static gboolean
+string_matcher_match_pcre(StringMatcher *self, const char *string, gsize string_len)
+{
+  gint rc;
+  gint num_matches;
+
+  if (pcre_fullinfo(self->pcre, self->pcre_extra, PCRE_INFO_CAPTURECOUNT, &num_matches) < 0)
+    g_assert_not_reached();
+  if (num_matches > RE_MAX_MATCHES)
+    num_matches = RE_MAX_MATCHES;
+
+  gsize matches_size = 3 * (num_matches + 1);
+  gint *matches = g_alloca(matches_size * sizeof(gint));
+
+  rc = pcre_exec(self->pcre, self->pcre_extra, string, string_len, 0, 0, matches, matches_size);
+  if (rc == PCRE_ERROR_NOMATCH)
+    {
+      return FALSE;
+    }
+  if (rc < 0)
+    {
+      msg_error("Error while matching pcrep", evt_tag_int("error_code", rc));
+      return FALSE;
+    }
+  if (rc == 0)
+    {
+      msg_error("Error while storing matching substrings");
+      return FALSE;
+    }
+  return TRUE;
+}
+
+static gboolean
+string_matcher_match(StringMatcher *self, const char *string, gsize string_len)
+{
+  switch (self->mode)
+    {
+    case SMM_LITERAL:
+      return (strcmp(string, self->pattern) == 0);
+    case SMM_PREFIX:
+      return (strncmp(string, self->pattern, strlen(self->pattern)) == 0);
+    case SMM_SUBSTRING:
+      return (strstr(string, self->pattern) != NULL);
+    case SMM_GLOB:
+      return (g_pattern_match_string(self->glob, string));
+    case SMM_PCRE:
+      return (string_matcher_match_pcre(self, string, string_len));
+    default:
+      g_assert_not_reached();
+    }
+}
+
+static StringMatcher *
+string_matcher_new(StringMatchMode mode, const gchar *pattern)
+{
+  StringMatcher *self = g_new0(StringMatcher, 1);
+
+  self->mode = mode;
+  self->pattern = g_strdup(pattern);
+
+  return self;
+}
+
+static void
+string_matcher_free(StringMatcher *self)
+{
+  if (self->pattern)
+    g_free(self->pattern);
+  if (self->glob)
+    g_pattern_spec_free(self->glob);
+  if (self->pcre)
+    pcre_free(self->pcre);
+  if (self->pcre_extra)
+    pcre_free_study(self->pcre_extra);
+  g_free(self);
+}
+
+typedef struct _ListSearchState
+{
+  TFSimpleFuncState super;
+  StringMatcher *matcher;
+  gint start_index;
+} ListSearchState;
+
+static void
+list_search_state_free(gpointer s)
+{
+  ListSearchState *self = (ListSearchState *)s;
+
+  if (self->matcher)
+    string_matcher_free(self->matcher);
+  tf_simple_func_free_state(&self->super);
+}
+
+static gboolean
+_list_search_mode_str_to_string_match_mode(const gchar *mode_str, StringMatchMode *string_match_mode)
+{
+  gboolean result = TRUE;
+
+  if (mode_str == NULL || strcmp(mode_str, "literal") == 0)
+    *string_match_mode = SMM_LITERAL;
+  else if (strcmp(mode_str, "prefix") == 0)
+    *string_match_mode = SMM_PREFIX;
+  else if (strcmp(mode_str, "substring") == 0)
+    *string_match_mode = SMM_SUBSTRING;
+  else if (strcmp(mode_str, "glob") == 0)
+    *string_match_mode = SMM_GLOB;
+  else if (strcmp(mode_str, "pcre") == 0)
+    *string_match_mode = SMM_PCRE;
+  else
```

Comment:
this functionality (to map the string to the SMM_ enum), could be moved alongside the StringMatcher interface as it could be useful for other uses of the interface.

Reply:
Thanks for the tip, as I mentioned in the other comment, I am preparing another PR for this :)